src/Workspaces/Core/Portable/Recommendations/AbstractRecommendationServiceRunner.cs

```diff
@@ -123,19 +117,38 @@ private ImmutableArray<ITypeSymbol> GetTypeSymbols(ImmutableArray<ISymbol> candi
                             type = allTypeArguments[0];
                         }
 
-                        // If type is <see cref="Func{TSomeArguments, TResult}"/>, extract TSomeArguments 
-                        // and find the type of thee argument corresponding to the ordinal.
-                        if (type is INamedTypeSymbol funcSymbolNamedTypeCandidate &&
-                            Equals(funcSymbolNamedTypeCandidate.Name, funcSymbol.Name) &&
-                            Equals(funcSymbolNamedTypeCandidate.ContainingNamespace, funcSymbol.ContainingNamespace))
+                        if (type.IsDelegateType())
                         {
                             var allTypeArguments = type.GetAllTypeArguments();
-                            if (allTypeArguments.Length < ordinalInLambda)
+
+                            // Actual <see cref="Delegate"/>
+                            if (allTypeArguments.Length == 0)
                             {
-                                continue;
+                                var methods = type.GetMembers(WellKnownMemberNames.DelegateInvokeName);
+                                if (methods.Length != 1)
+                                {
+                                    continue;
+                                }
+
+                                var parameters = methods[0].GetParameters();
+                                if (parameters.Length < ordinalInLambda)
+                                {
+                                    continue;
+                                }
+
+                                type = parameters[ordinalInLambda].Type;
                             }
+                            // Functions, Actions and other delegates with TypeArguments.
+                            else
```

Comment:
You don't need this else branch at all: you should *only* be looking at the Invoke method. Consider the case too like:

`delegate void ReversingArguments<T1, T2>(T2 t2, T1 t1)`

The Invoke method will correctly have the order; the type argument order isn't right.

Reply:
Or just something like [`EventHandler<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.eventhandler-1) where the generic argument isn't the first argument.

Reply:
Thank you, @jasonmalinowski ! Added a test for this, and moved all the code to use Invoke.

---
In reply to: [276816954](https://github.com/dotnet/roslyn/pull/35067#discussion_r276816954) [](ancestors = 276816954)

Reply:
Makes sense. Thank you! I think that the reversing argument test covers the switch in code.

---
In reply to: [276817775](https://github.com/dotnet/roslyn/pull/35067#discussion_r276817775) [](ancestors = 276817775)